### PR TITLE
FIFO the Datalog!

### PR DESCRIPTION
**Pull request by adamemerson**
Set the RGW Data Changes Log atop CLS FIFO.

(My original plan was to do this and the Metadata Log by changing RGWSI_Cls::TimeLog out from under them, but since the Metadata Log seems to want to trim by time, it can have a pass of its own.)

### CODE REVIEW

**Comment by mattbenjamin** [comment]
@adamemerson is this complete?

**Comment by adamemerson** [comment]
@mattbenjamin Needs testing and a rebase, but yes. Though I want to rip out the dedicated thread and use an Asio timer.

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -8582,12 +8582,12 @@ int main(int argc, const char **argv)
     RGWDataChangesLog::LogMarker log_marker;
 
     do {
-      list<rgw_data_change_log_entry> entries;
+      std::vector<rgw_data_change_log_entry> entries;
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc:8585]: yay

File: src/rgw/rgw_trim_datalog.cc [hunk 2]
```diff
@@ -128,9 +128,13 @@ int DataLogTrimCR::operate()
         ldout(cct, 10) << "trimming log shard " << i
             << " at marker=" << m
             << " last_trim=" << last_trim[i] << dendl;
-        spawn(new TrimCR(store, store->svc()->datalog_rados->get_oid(i),
-                         m, &last_trim[i]),
-              true);
+#warning BOOM!
```
Comment (mattbenjamin) [inline src/rgw/rgw_trim_datalog.cc]: boom?

File: src/rgw/rgw_trim_datalog.cc [hunk 3]
```diff
@@ -160,7 +164,10 @@ class DataLogTrimPollCR : public RGWCoroutine {
                     int num_shards, utime_t interval)
     : RGWCoroutine(store->ctx()), store(store), http(http),
       num_shards(num_shards), interval(interval),
-      lock_oid(store->svc()->datalog_rados->get_oid(0)),
+
+#warning BOOM!
```
Comment (mattbenjamin) [inline src/rgw/rgw_trim_datalog.cc]: ib. boom

File: src/rgw/rgw_trim_datalog.cc [hunk 4]
```diff
@@ -160,7 +164,10 @@ class DataLogTrimPollCR : public RGWCoroutine {
...
+      lock_oid(string{}),
+//      lock_oid(store->svc()->datalog_rados->get_oid(0)),
```
Comment (mattbenjamin) [inline src/rgw/rgw_trim_datalog.cc]: ?

**Comment by mattbenjamin** [comment]
what about N backports?  (w/o neorados)

File: src/rgw/rgw_trim_datalog.cc (hunk 2 above)
Comment (adamemerson) [reply src/rgw/rgw_trim_datalog.cc]: This gets fixed in a later commit.

**Comment by adamemerson** [comment]
@mattbenjamin I'll get on the N backport next thing.

**Comment by adamemerson** [comment]
@mattbenjamin Also I just realize I need a configuration setting to allow people to select the original version. I'll shove that in next thing.

File: src/rgw/rgw_trim_datalog.cc (hunk 2 above)
Comment (markhpc) [reply src/rgw/rgw_trim_datalog.cc]: I love comments like this, definitely should keep! :smile:

**Pull request closed without merging**